File: DP/neetcode-practise/coin-change-2.js
function numberOfWays(target, arr, i, sum, dp = {}) {
  //console.log(i," ",sum);
  if (sum === target) {
    // console.log(dp);
    return 1;
  }
  if (sum > target || i >= arr.length) {
    return 0;
  }
  if (dp[`${i},${sum}`]) {
    return dp[`${i},${sum}`];
  }
  let newSum = sum + arr[i];
  dp[`${i},${newSum}`] = numberOfWays(target, arr, i, newSum, dp)+numberOfWays(target, arr, i + 1, sum, dp);
  return dp[`${i},${newSum}`];
}

let arr = [1, 2, 5];
let sum = 500;
console.log(numberOfWays(sum, arr, 0, 0));

File: DP/neetcode-practise/coin-change2.py
class Solution:
    def change(self, amount, coins):
        dp = [0 * (len(coins) + 1) for i in range(amount + 1)]
        dp[0] = [1] * (len(coins) + 1)

        for a in range(1, amount + 1):
            for w in range(len(coins) - 1, -1, -1):
                dp[a][w] = dp[a][w + 1]
                if a - coins[w] >= 0:
                    dp[a][w] += dp[a - coins[w]][w]
        return dp[amount][0]


arr = [1, 2, 5]
Sum = 5
print(Solution().change(Sum, arr))
